### DISCUSSION TITLE
Canceling long-living effects

### DISCUSSION THREAD

**Discussion by artem-krachulov-paysera** [pointfreeco/swift-composable-architecture#1923] [General]
After more than one year of developing an app with SwiftUI and TCA I use one rule of canceling effects for optional sores or switching stores.

Rule:
1. First cancel effects for target reducer, usually from parent reducer (by calling child cancel action which cancels all ids)
2. Set child state to nil or switch with enum

App grows, views wrap to another view, logic gets more complicated, and releasing view states must be triggered from different reducers and places. Managing canceling effects becomes pain which produces leaks.

Is there a way to use `.onDissapear{}` modifier and cancel all effects one view disappears.

**Comment by tgrapperon** [comment]
Hey @artemkrachulov. Indeed, cancelling from a Parent doesn't scale well, but this is more or less the only reliable way right now, given that `onDisappear` is only called when the complete feature did effectively disappear, and not as soon as its state is `nil`'d.

You can sometimes tie your `Effect`'s lifetime by awaiting the `ViewStoreTask` like it is done in the "Long-living Effect" case study, but this is not always possible. But in these cases, the child feature can handle creation and cancellation of the `Effect` locally.

Please note that the upcoming `navigation` feature will automatically handle cancellation of `Effect`s when the views are going away.
